runtime/compiler/trj9/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -15665,10 +15665,60 @@ J9::Z::TreeEvaluator::BCDCHKEvaluatorImpl(TR::Node * node,
                                           bool isUseVector)
    {
    TR_Debug* debugObj = cg->getDebug();
-   TR::Node * secondChild = node->getChild(1);
+   TR::Node* secondChild = node->getChild(1);
+   TR::Compilation* comp = cg->comp();
+   TR::Register* bcdOpResultReg = NULL;
+
    bool isVariableParam = pdopNode->getOpCodeValue() == TR::lcall ||
                           pdopNode->getOpCodeValue() == TR::icall;
 
+   // Skip the evaluation of pdshlOverflow if its child i2pd (or l2pd) is already evaluated.
+   //
+   // If vector instructions are used, the pdshlOverflow+i2pd structure only requires one VCVD(G) instruction to achieve
+   // the desired functionality. Effectively, the pdshlOverflow does not need any evaluation. The value (register) associated with
+   // i2pd can be copied over to the pdshlOverflow node. This also implies that the
+   // pdshlOverflow node is not an exceptioning node. Thus, generating an OOL sequence for a non-exceptioning node no only
+   // violates the definition of BCDCHK node, but also can make RA generate bad GC map and eventually cause GC error during runtime
+   //
+   if (isUseVector)
+      {
+      bool hasEvaluatedI2PD = pdopNode->getOpCodeValue() == TR::pdshlOverflow
+              && (pdopNode->getFirstChild()->getOpCodeValue() == TR::i2pd ||
+                  pdopNode->getFirstChild()->getOpCodeValue() == TR::l2pd)
+              && pdopNode->getFirstChild()->getRegister() != NULL;
+
+      if (hasEvaluatedI2PD && pdopNode->getRegister() == NULL)
```

Comment:
Why is `pdshlOverflows` child relevant here? Do we not only have to check that the `pdshlOverflow` itself has been evaluated or not?

Reply:
The case you mentioned (where `BCDCHK` node's first child has been evaluated)  has been covered at another place. See `J9::Z::TreeEvaluator::BCDCHKEvaluator(TR::Node * node, TR::CodeGenerator * cg)`

This particular change is to cover a particular corner case, where a `pdshlOverflow` is not evaluated but its child `i2pd` is. If this is the case, this `pdshlOverflow` won't evaluate to anything in the mainline. We then end up having an empty mainline and a non-empty OOL. 

Essentially, the codegen should avoid evaluating `BCDCHK` nodes whose EXCEPTIONING node has been evaluated. And on z14, the `i2pd` [under `pdshlOverflow`, which is in turn under a `BCDCHK`] becomes the exceptioning node. This corner case slightly violates the definition of a `BCDCHK` tree, where the first child should be the exceptioning child. This is why I thought about generating different IL for z14 and non-z14 targets. That would prevent the codegen from dealing different corner cases like this one we are looking at.

Reply:
> The case you mentioned (where BCDCHK node's first child has been evaluated) has been covered at another place. See J9::Z::TreeEvaluator::BCDCHKEvaluator(TR::Node * node, TR::CodeGenerator * cg)

This is a bug then. We cannot skip evaluation of the entire `BCDCHK`. What if one of the call children which hold a symref are commoned? We would effectively skip the evaluation of node with a symref. We need to fix this. Might as well do it in this PR. Can you take a look?

> This particular change is to cover a particular corner case, where a pdshlOverflow is not evaluated but its child i2pd is. If this is the case, this pdshlOverflow won't evaluate to anything in the mainline. We then end up having an empty mainline and a non-empty OOL.

This may not be true. The two cases are:

1. `i2pd` is commoned

In this case it must have been commoned with another `i2pd` previously with precision 10, which means it cannot overflow. The `pdshlOverfow` node in this case is a truncation operation which must mean a `VSRP` must be generated.

2. `i2pd` is not commoned

In this case we can do a fast path of generating `VSRP` if and only if the `i2pd` has a reference count of 1, because we will effectively be clobbering the evaulation using `VCVD`.

In both cases above we generate at least one instruction, meaning the mainline is non-empty. Can we address these concerns?